src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProject.cs

```diff
@@ -473,6 +503,77 @@ public void RemoveAdditionalFile(string fullPath)
 
         #endregion
 
+        #region Non Source File Addition/Removal
+        public void AddDynamicSourceFile(string dynamicFilePath, ImmutableArray<string> folders)
+        {
+            var extension = FileNameUtilities.GetExtension(dynamicFilePath)?.TrimStart('.');
+            if (extension?.Length == 0)
+            {
+                return;
+            }
+
+            foreach (var provider in _dynamicFileInfoProviders)
+            {
+                // skip unrelated providers
+                if (!provider.Metadata.Extensions.Any(e => string.Equals(e, extension, StringComparison.OrdinalIgnoreCase)))
+                {
+                    continue;
+                }
+
+                // don't get confused by _filePath and filePath
+                // VisualStudioProject._filePath points to csproj/vbproj of the project
+                // and the parameter filePath points to dynamic file such as cshtml and etc
+                // 
+                // also, provider is free-threaded. so find to call Wait rather than JTF
+                var fileInfo = provider.Value.GetDynamicFileInfoAsync(
+                    projectId: Id, projectFilePath: _filePath, filePath: dynamicFilePath, CancellationToken.None).WaitAndGetResult_CanCallOnBackground(CancellationToken.None);
+
+                if (fileInfo == null)
+                {
+                    continue;
+                }
+
+                fileInfo = FixUpDynamicFileInfo(fileInfo, dynamicFilePath);
+
+                // remember map between original dynamic file path to DynamicFileInfo.FilePath
+                _dynamicFilePathMaps.Add(dynamicFilePath, fileInfo.FilePath);
```

Comment:
.Add will throw if we get teh same path. we probably want to be checking for this up front and enforcing invariants about how we're used.

Reply:
not in this PR. but, if we believe that is important, we can open an issue to track it. if people believe that's important, someone will clean that up.

Reply:
Could you explain why it's not important htat we check and enforce invariants as part of introducing a new PR?  You're effectively shifting hte burdern onto others who already have existing work to do.  I would expect it to be part of any large/new feature work, or any new subsystem/capability that we're at least checking our args.  Otherwise, this doesn't feel to me like this meets minimum standards to go into IDE/workspace.